py-rse/git-advanced.Rmd

```diff
@@ -116,28 +181,253 @@ We can (and should) choose the name to signal the purpose of the branch,
+
+> "The most frequent word will occur approximately \(2^\alpha\) times
+> as often as the second most frequent word,
+> \(3^\alpha\) times as often as the third most frequent word, and so on."
+
+This mathematical expression for Zipf's Law is an example of a power law.
+In general, when two variables \(x\) and \(y\)
+are related through a power law, \[
+y = ax^b
+\]
+
+taking logarithms of both sides yields a linear relationship: \[
+\log(y) = \log(a) + b\log(x)
+\]
+
+Hence, plotting the variables on a log-log scale reveals this
+linear relationship.
+
+#### Estimating alpha
+
+In order to determine the best method for estimating \(\alpha\) and fitting a line to our data,
+we turn to a scientific paper on Zipf’s Law in english texts [@Moreno2016].
+The mathematics in the paper is a little heavy,
+but the authors essentially suggest using a method called maximum likelihood estimation. 
+This involves defining a so-called negative log likelihood function,
+whose variables include the word count data and a \(\beta\) parameter.
+As we'll see below,
+this \(\beta\) parameter is related to the \(\alpha\) parameter in our definition of Zipf's Law
+as well as the \(b\) parameter in the general form of a power law.
+This means we can use it to verify Zipf's Law *and* plot our fitted curve. 
+Expressed as a python function, the likelihood function reads as follows:
+
+```python
+import numpy as np
+
+def nlog_likelihood(beta, counts):
+    """Log-likelihood function."""
+    likelihood = - np.sum(np.log((1/counts)**(beta - 1) - (1/(counts + 1))**(beta - 1)))
+    return likelihood
+``` 
+
+Obtaining an estimate of the \(\beta\) (and thus \(\alpha\)) parameter then becomes
+a mathematical optimization problem,
+for which we can use the `scipy.optimize` library.
+Following @Moreno2016,
+we use Brent's method with \(1 < \beta \leq 4\).
+
+```python
+from scipy.optimize import minimize_scalar
+
+def get_power_law_params(word_counts):
+    """Get the power law parameters."""
+    mle = minimize_scalar(nlog_likelihood, bracket=(1 + 1e-10, 4),
+                          args=(word_counts), method='brent')
+    beta = mle.x
+    alpha = 1 / (beta - 1)
+    return alpha, beta
+```
+
+We can then plot the fitted curve on the plot axes (`ax`)
+defined in the `plotcounts.py` script,
+
+```python
+def plot_fit(curve_xmin, curve_xmax, max_rank, beta, ax):
+    """
+    Plot the power law curve that was fitted to the data.
+
+    Parameters
+    ----------
+    curve_xmin : float
+        Minimum x-bound for fitted curve
+    curve_xmax : float
+        Maximum x-bound for fitted curve
+    max_rank : int
+        Maximum word frequency rank.
+    beta : float
+        Estimated beta parameter for the power law.
+    ax : matplotlib axes
+        Scatter plot to which the power curve will be added.
+    """
+    xvals = np.arange(curve_xmin, curve_xmax)
+    yvals = max_rank * (xvals**(-beta + 1))
```

Comment:
Maybe we should rewrite this to use `alpha` not `beta`, so it links better with the previous section.

Reply:
That's probably a good idea. I guess we can wait until you resolve #367 to rewrite, since the rewrite also needs to acknowledge the variable C which way say we'll refer back to but don't as yet.